### ISSUE TITLE
Retrieve prefixes by class instead of string in SystemOfUnitsService

### ISSUE THREAD

**Issue by keilw** [unitsofmeasurement/unit-api#83]
Currently SystemOfUnitsService returns `Collection<Prefix> getPrefixes(String name);` as of https://github.com/unitsofmeasurement/unit-api/issues/81.

Let's try to retrieve it using the enum class (e.g. MetricPrefix) directly, so something like `return Collections.<Prefix>unmodifiableSet(EnumSet.allOf(MetricPrefix.class));` would take the `Class` of a particular `Prefix `implementation. If we used a `default` method in the API, every `Prefix` should be based on enums, otherwise we may leave that to implementations like Indriya.

**Comment by desruisseaux** [comment]
Not sure why we need to specify a class? If the `SystemOfUnitsService` interface is implemented by libraries, it seems to me that those libraries known their implementation class?

If for some reason we really want to keep the `Class` argument, then at the very least it shall be `Class<?>` (or more specific type is we know it) and the `@SuppressWarnings("rawtypes")` should be removed.

**Comment by desruisseaux** [comment]
It is not the only way it will work. The clean way to work with `Class<?> prefixType` is:

```
EnumSet.allOf(prefixType.asSubclass(Enum.class))
```

But even if for some reason you wanted raw types, keep it in the implementation, not in the API. You can have `Class<?>` in the public interface and only `Class` in the implementing method (because the JVM does type erasure as you mentioned).

**Comment by desruisseaux** [comment]
However the above does not answer the question about why we need a `Class<?>` argument in the first place.

**Comment by keilw** [comment]
It will not work with Enum.class that isn't specific enough, you have to use
```
EnumSet.allOf(MetricPrefix.class))
````
And the Class instead of a String as argument avoids boilerplate code. 

If you made the `AsSubclass` option work and compile in the RI, please push or PR it there.

**Comment by desruisseaux** [comment]
Werner, it will work with `Enum.class`. Read the Javadoc.

**Comment by keilw** [comment]
If you have time to propose a PR or push it directly, please feel free to do. I am not on my own computer but maybe tonight or on the weekend I could also do. 

If trying to apply `asSubclass(Enum.class)` would also throw an appropriate exception (ClassCastException or similar) when you try to pass say `Number.class` instead of `MetricPrefix.class` then it would even make it a one-liner ;-)

**Comment by desruisseaux** [comment]
Indeed, `prefixType.asSubclass(Enum.class)` throws `ClassCastException` if `prefixType` is not a subtype of `Enum`. I will try to create pull requests soon.

**Comment by keilw** [comment]
Merged everything. Thanks @desruisseaux, I also removed the extra if/else check and added a JUnit test for `ClassCastException`. I added author references and mentioned it may throw a `ClassCastException` in the API. Note, it is implementation-specific. A different implementation may also chose a Map of Sets with the Class as a key to retrieve the content of a `Prefix` implementation. Using enum keeps it flexible and it works even for unknown implementations (in fact it should work even for those declared by another implementation here as long as it also relies on enum ;-) but it is not a requirement of the Spec or API that enums must be used.

**Comment by desruisseaux** [comment]
Thanks. But it still not clear to me how the `Class<?>` argument achieve implementation independence? In the way it is currently done, isn't the user have to know the `Enum` type (s)he is using, which is specific to every implementation?

**Comment by keilw** [comment]
The class has to be from a particular implementation, but at least those who use enums that implementation does not even have to be Indriya, it could also be UOMo or SIS (should Prefix even be relevant there)

**Comment by desruisseaux** [comment]
But still, the user can not said "give me the SI prefix of whatever implementation is behind that `SystemOfUnitsService` implementation". He become tied to a particular implementation.

For a library for example, it become impossible to let the user choose its JSR 385 implementation and get the SI prefixes of that implementation. The library would have to said for example `getPrefixes(TheIndriyaImplementationOfPrefix.class)`, which is equivalent to forcing the library user to choose that particular implementation.

**Comment by desruisseaux** [comment]
I understand that any implementation based on `Class.getEnumConstants()` can work with any `Class<? extends Enum>`, but a user who want to use this API still have to know the `TheUOMoImplementationOfPrefix` class. He has to tie himself to a particular implementation.

A `String` argument would avoid this problem, provided that we standardize some `String` values (e.g. `"SI"`). I would be fine with other type of argument too, as long as they allow the use of the API in an implementation-independent way. Current use of the `Class` argument does not allow that (even if Indriya can get "foreigner" `Enum`, the user is still tied to a particular implementation).

**Comment by keilw** [comment]
It's actually a little more complicated than simply chosing class or string. I created https://github.com/unitsofmeasurement/unit-api/issues/85, because the current approach in any case forces **EVERY** module to override that method even if it has no use for prefixes. E.g. SI uses the same prefixes we declared in the RI already, In theory if another implementation decides to only go with BinaryPrefix or another prefix it could declare them in another module. `MetricPrefix` is so common it makes sense in the RI. So we have to decide if we go for a simple one-liner possibly even as default method in the API works for most users (it still allows to override for every implementation that needs to store it a different way) or add the same boilerplate we do for unit systems, but unless we decided to introduce a totall new `PrefixService` just for the prefix, it must be implemented in an abstract way so all systems can avoid it and reuse from the implementation if they do not need their own prefix.
